apm-agent-plugins/apm-profiling-plugin/src/main/java/co/elastic/apm/agent/profiler/SamplingProfiler.java

```diff
@@ -0,0 +1,287 @@
+        }
+    }
+
+    public void onDeactivation(Thread thread, TraceContextHolder<?> activeSpan, @Nullable TraceContextHolder<?> previouslyActive) {
+        if (profilingSessionOngoing) {
+            activationEvents.offer(ActivationEvent.deactivation(activeSpan, thread.getId(), previouslyActive));
+        }
+    }
+
+    CallTree.Root getRoot() {
+        return profiledThreads.get(Thread.currentThread().getId());
+    }
+
+    void setProfilingSessionOngoing(boolean profilingSessionOngoing) {
+        this.profilingSessionOngoing = profilingSessionOngoing;
+    }
+
+    void clear() {
+        profiledThreads.clear();
+        activationEvents.clear();
+    }
+
+    @Override
+    public void run() {
+        TimeDuration sampleRate = config.getSampleRate();
+        TimeDuration profilingDuration = config.getProfilingDuration();
+        TimeDuration profilingInterval = config.getProfilingInterval();
+        profilingSessionOngoing = true;
+
+        logger.debug("Start profiling session");
+        profile(sampleRate, profilingDuration);
+        logger.debug("End profiling session");
+
+        if (profilingInterval.getMillis() != 0) {
+            profilingSessionOngoing = false;
+            // TODO do we want to create inferred spans for partially profiled transactions?
+            profiledThreads.clear();
+        }
+        // clears the interrupted status so that the thread can return to the pool
+        if (!Thread.interrupted()) {
+            long delay = profilingInterval.getMillis() - config.getProfilingDuration().getMillis();
+            scheduler.schedule(this, delay, TimeUnit.MILLISECONDS);
+        }
+    }
+
+    private void profile(TimeDuration sampleRate, TimeDuration profilingDuration) {
+        long sampleRateMs = sampleRate.getMillis();
+        long deadline = profilingDuration.getMillis() + System.currentTimeMillis();
+        while (!Thread.currentThread().isInterrupted() && System.currentTimeMillis() < deadline) {
+            try {
+                long startNs = System.nanoTime();
+
+                takeThreadSnapshot();
+
+                long durationMs = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startNs);
+                logger.trace("Taking snapshot took {}ms", durationMs);
+
+                Thread.sleep(Math.max(sampleRateMs - durationMs, 0));
+            } catch (RuntimeException e) {
+                logger.error("Exception while taking profiling snapshot", e);
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+                break;
+            }
+        }
+    }
+
+    private void takeThreadSnapshot() {
+        List<WildcardMatcher> excludedClasses = config.getExcludedClasses();
+        List<WildcardMatcher> includedClasses = config.getIncludedClasses();
+        long nanoTime = System.nanoTime();
+        processActivationEventsUpTo(nanoTime);
+        long[] profiledThreadIds = getProfiledThreadIds();
```

Comment:
What if, instead of only thread IDs, you would use a different method that returns a map between thread ID and the active span. Then, right after you get the stack traces, you check again and you only use the stack trace of a thread if the current active span is still the same as before the capturing and if not- you discard it altogether (which is not so bad, as if there was only one such it is not interesting anyway). Can this fix the problem of wrong child-parent inference?

Reply:
Are you suggesting to add inferred spans to the activation stack? That doesn't work if we want to integrate with async-profiler, for example. That's because we'd record all activity within a profiling session and only afterwards process activation events and stack traces. Also, we don't know which CallTree nodes to convert to a span until they have ended.

> To reduce the number of spans we create, we only create spans for CallTree nodes which have appeared in at least two consecutive samples (count >= 2). Also, if a node has only one child and that child has the same count we don’t create a span for it but instead add it to the span.stacktrace of the leaf node.

Reply:
> Are you suggesting to add inferred spans to the activation stack?

No, I am suggesting a way to make sure you know which is the actual parent of the inferred span. If the same (real) span is active before and after you capture the stack trace, you can tell it is a real parent of anything you infer from this stack trace. If the active span changes while you took the stack trace, you can decide to discard this sample.

Reply:
I think I see what you mean but that's not what causes the parent/child relationship mismatches. I've updated the `Caveats` session of https://github.com/elastic/apm-agent-java/issues/971#issuecomment-566054290 to explain the scenarios which leat to a mismatch.

Reply:
Ahh, I see.